pkgs/build-support/docker/default.nix

```diff
@@ -994,4 +996,177 @@ rec {
+          group = ''
+            root:x:0:
+            nixbld:!:${toString gid}:
+            nogroup:x:65534:
+          '';
+          passAsFile = [ "passwd" "group" ];
+        } ''
+          # Directories that are later made writable in fakeRootCommands
+          mkdir -p "$out"{/tmp,${storeDir},${sandboxBuildDir}}
+
+          # Directories for internet access and POSIX compat
+          mkdir -p "$out"{/etc/ssl/certs,/bin,/usr/bin}
+          # Allows Nix binaries to find CA certificates for internet access
+          ln -s ${cacert}/etc/ssl/certs/ca-bundle.crt "$out"/etc/ssl/certs/ca-certificates.crt
+          # POSIX compat, for shell scripts and co.
+          ln -s ${bashInteractive}/bin/sh "$out"/bin/sh
+          ln -s ${coreutils}/bin/env "$out"/usr/bin/env
+          # Allows programs to look up the build users home directory
+          cp "$passwdPath" "$out"/etc/passwd
+          cp "$groupPath" "$out"/etc/group
+        '';
+
+        # This function closely mirrors what this Nix code does:
+        # https://github.com/NixOS/nix/blob/2.8.0/src/libexpr/primops.cc#L1102
+        # https://github.com/NixOS/nix/blob/2.8.0/src/libexpr/eval.cc#L1981-L2036
+        stringValue = value:
+          # We can't just use `toString` on all derivation attributes because that
+          # would not put path literals in the closure. So we explicitly copy
+          # those into the store here
+          if builtins.typeOf value == "path" then "${value}"
+          else if builtins.typeOf value == "list" then toString (map stringValue value)
+          else toString value;
+
+        # https://github.com/NixOS/nix/blob/2.8.0/src/libstore/build/local-derivation-goal.cc#L992-L1004
+        drvEnv = lib.mapAttrs' (name: value:
+          let str = stringValue value;
+          in if lib.elem name (drv.drvAttrs.passAsFile or [])
+          then lib.nameValuePair "${name}Path" (writeText "pass-as-text-${name}" str)
+          else lib.nameValuePair name str
+        ) drv.drvAttrs //
+          # A mapping from output name to the nix store path where they should end up
+          # https://github.com/NixOS/nix/blob/2.8.0/src/libexpr/primops.cc#L1253
+          lib.genAttrs drv.outputs (output: builtins.unsafeDiscardStringContext drv.${output}.outPath);
+
+        # Environment variables set in the image
+        envVars = {
+
+          # https://github.com/NixOS/nix/blob/2.8.0/src/libstore/build/local-derivation-goal.cc#L1027-L1030
+          PATH = "/path-not-set";
+
+          # https://github.com/NixOS/nix/blob/2.8.0/src/libstore/build/local-derivation-goal.cc#L1032-L1038
+          HOME = "/homeless-shelter";
+
+          # https://github.com/NixOS/nix/blob/2.8.0/src/libstore/build/local-derivation-goal.cc#L1040-L1044
+          NIX_STORE = storeDir;
+
+          # https://github.com/NixOS/nix/blob/2.8.0/src/libstore/build/local-derivation-goal.cc#L1046-L1047
+          # TODO: Make configurable?
+          NIX_BUILD_CORES = "1";
+
```

Comment:
I'd love if Nix would be able to automatically detect this, based on the number of CPUs it sees.

For the time being, It's probably fine to set this to `1` - but the docs should document this can be overridden if more cores should be used.

Reply:
Isn't `auto` a valid value?

Reply:
Oh I guess that's only on the config side. This is supposed to be the script-friendly side of it.

Reply:
It's ok to leave it to the user commands. We may want to "cheat" a bit and match users expectations when they set this environment variable in the image config `Env`/`Environment` or what it is. So only set it if missing?

Reply:
Users can't directly customize the `Env`, but they can set an environment variable as a derivation attribute, which would do the job. I think this currently works all as expected.